gradle-plugin-testing-junit/src/main/java/com/palantir/gradle/testing/junit/GradleProjectParameterResolver.java

```diff
@@ -18,38 +18,48 @@
 
 import com.palantir.gradle.testing.project.RootProject;
 import com.palantir.gradle.testing.project.SubProject;
+import java.util.Arrays;
 import java.util.Optional;
+import org.junit.jupiter.api.TestTemplate;
 import org.junit.jupiter.api.extension.ExtensionContext;
 import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.platform.commons.annotation.Testable;
 
 final class GradleProjectParameterResolver implements TerseParameterResolver {
     @Override
     public Optional<Object> parameter(ParameterContext parameterContext, ExtensionContext extensionContext) {
         if (parameterContext.getParameter().getType().equals(RootProject.class)) {
-            return Optional.of(rootProjectFor(parameterContext, extensionContext));
+            return Optional.of(rootProjectFor(extensionContext));
         }
 
         if (parameterContext.getParameter().getType().equals(SubProject.class)) {
-            return Optional.of(rootProjectFor(parameterContext, extensionContext)
-                    .subproject(projectNameFromParameter(parameterContext)));
+            if (!isTestMethod(parameterContext)) {
+                throw new IllegalStateException("""
+                    SubProject parameters can only be injected in test methods \
+                    (@Test, @ParameterizedTest, @RepeatedTest, @TestFactory, etc.), \
+                    not in lifecycle methods like @BeforeEach or @AfterEach. \
+                    Use RootProject.subproject("name") explicitly in lifecycle methods instead. \
+                    Found SubProject parameter '%s' in method '%s'\
+                    """.formatted(
+                                parameterContext.getParameter().getName(),
+                                parameterContext.getDeclaringExecutable().getName()));
+            }
```

Comment:
I don't really think we need this check; at least if we're getting people to understand that `SubProject foo` makes a subproject called `foo`, it's not too bad to let people use `SubProject` in `beforeEach`.